### PR TITLE
Update mapbox sdk 7.2.0

### LINKED ISSUE #19: Add Offline switch on ZEIR
Just add a switch for offline maps mode which downloads the Maps when the mode is activated

### PR DESCRIPTION
**Pull request by ekigamba**
Fixes #19

### CODE REVIEW

**Comment by githengi** [comment]
Update the ReadMe and other documentation

File: utils/src/test/java/io/ona/kujaku/helpers/converters/GeoJSONHelperTest.java [hunk 1]
```diff
@@ -190,7 +190,7 @@ public void getFeatureShouldReturnMultiPointFeatureWithProperties() throws JSONE
     private LatLng generateLatLng() {
         return new LatLng(
                 (Math.random() * (GeometryConstants.MAX_LATITUDE - GeometryConstants.MIN_LATITUDE)) + GeometryConstants.MIN_LATITUDE,
-                (Math.random() * (GeometryConstants.MAX_LONGITUDE - GeometryConstants.MIN_LONGITUDE)) + GeometryConstants.MIN_LONGITUDE
+                (Math.random() * (170d - (-170d))) + GeometryConstants.MIN_LONGITUDE
```
Comment (vincent-karuri) [inline utils/src/test/java/io/ona/kujaku/helpers/converters/GeoJSONHelperTest.java]: why not just make this `* (340d)`?

File: library/src/main/java/io/ona/kujaku/views/KujakuMapView.java [hunk 2]
```diff
@@ -585,9 +585,7 @@ private void afterStyleLoadedOperations(@NonNull Style style) {
 
         addWmtsLayers();
 
-        if (PermissionsManager.areLocationPermissionsGranted(getContext())) {
-            mapboxLocationComponentWrapper.init(KujakuMapView.this.mapboxMap, getContext());
-        }
+        mapboxLocationComponentWrapper.init(KujakuMapView.this.mapboxMap, getContext());
```
Comment (vincent-karuri) [inline library/src/main/java/io/ona/kujaku/views/KujakuMapView.java]: Why remove this check?

It should only be initialized if permissions are set.

https://docs.mapbox.com/android/maps/overview/location-component/#activating

File: library/src/test/java/io/ona/kujaku/layers/ArrowLineLayerTest.java [hunk 3]
```diff
@@ -720,4 +721,21 @@ public void filterFeaturesByRegex() throws InvalidArrowLineConfigException {
         assertEquals(4, filteredFeatures.get(2).getNumberProperty("position"));
         assertEquals(5, filteredFeatures.get(3).getNumberProperty("position"));
     }
+
+    @Test
+    public void getLayerIds() throws InvalidArrowLineConfigException {
```
Comment (vincent-karuri) [inline library/src/test/java/io/ona/kujaku/layers/ArrowLineLayerTest.java]: Could you maybe use the convention where you start with the suffix `test` and say what to expect from the test (in the function name)?

File: library/src/test/java/io/ona/kujaku/layers/BoundaryLayerTest.java [hunk 4]
```diff
@@ -157,4 +158,28 @@ public void calculateCenterPoints() {
         assertPointEquals(Point.fromLngLat(6.5d, 6.5d), point1);
         assertPointEquals(Point.fromLngLat(9.1d, 9.1d), point2);
     }
+
+    @Test
+    public void getLayerIds() {
```
Comment (vincent-karuri) [inline library/src/test/java/io/ona/kujaku/layers/BoundaryLayerTest.java]: Same issue with the naming convention.

File: library/src/main/java/io/ona/kujaku/views/KujakuMapView.java [hunk 5]
```diff
@@ -1275,6 +1301,22 @@ public boolean changeLocationUpdates(long updateInterval, long fastestUpdateInte
...
+    public boolean isKujakuLayerAdded(@NonNull KujakuLayer kujakuLayer) {
+        String[] layerIds = kujakuLayer.getLayerIds();
+        if (mapboxMap != null && mapboxMap.getStyle() != null && mapboxMap.getStyle().isFullyLoaded()) {
+            for (String layerId: layerIds) {
+                if (mapboxMap.getStyle().getLayer(layerId) == null) {
+                    return false;
+                }
+            }
+
+            return true;
```
Comment (vincent-karuri) [inline library/src/main/java/io/ona/kujaku/views/KujakuMapView.java]: Can this be moved out of the conditional to be the default return value?

File: library/src/main/java/io/ona/kujaku/views/KujakuMapView.java [hunk 6]
```diff
@@ -1275,6 +1301,22 @@ public boolean changeLocationUpdates(long updateInterval, long fastestUpdateInte
         return false;
     }
 
+    @Override
+    public boolean isKujakuLayerAdded(@NonNull KujakuLayer kujakuLayer) {
+        String[] layerIds = kujakuLayer.getLayerIds();
+        if (mapboxMap != null && mapboxMap.getStyle() != null && mapboxMap.getStyle().isFullyLoaded()) {
+            for (String layerId: layerIds) {
```
Comment (vincent-karuri) [inline library/src/main/java/io/ona/kujaku/views/KujakuMapView.java]: Validate layerIds is not null?

File: sample/src/main/java/io/ona/kujaku/sample/activities/CaseRelationshipActivity.java [hunk 7]
```diff
@@ -81,27 +83,32 @@ public void onClick(View v) {
...
-
-                // Zoom to the position
-                mapboxMap.easeCamera(CameraUpdateFactory.newLatLngZoom(new LatLng(0.15380840901698828, 37.66387939453125), 8d));
+                mapboxMap.setStyle(Style.MAPBOX_STREETS, new Style.OnStyleLoaded() {
+                    @Override
+                    public void onStyleLoaded(@NonNull Style style) {
+                        GeoJsonSource sampleCasesSource = new GeoJsonSource(CASES_SOURCE_ID, sampleCases);
+                        style.addSource(sampleCasesSource);
+
+                        Expression colorExpression = match(get("testStatus")
```
Comment (vincent-karuri) [inline sample/src/main/java/io/ona/kujaku/sample/activities/CaseRelationshipActivity.java]:
```
 Expression colorExpression = match(get("testStatus")
                                , rgba(0, 0, 0, 0)
                                , stop(literal("positive"), Expression.color(getColorv16(R.color.positiveTasksColor)))
                                , stop(literal("negative"), Expression.color(getColorv16(R.color.negativeTasksColor))));

                        FillLayer fillLayer = new FillLayer("sample-cases-fill", CASES_SOURCE_ID);
                        fillLayer.withFilter(neq(geometryType(), "Point"));
                        fillLayer.withProperties(fillColor(colorExpression));

                        CircleLayer circleLayer = new CircleLayer("sample-cases-symbol", CASES_SOURCE_ID);
                        circleLayer.withFilter(eq(geometryType(), "Point"));
                        circleLayer.withProperties(circleColor(colorExpression), circleRadius(10f));

                        style.addLayer(circleLayer);
                        style.addLayer(fillLayer);
```

Could this move to a separate method?

File: library/src/main/java/io/ona/kujaku/views/KujakuMapView.java [hunk 8]
```diff
@@ -1069,10 +1091,7 @@ public void onResult(LocationSettingsResult result) {
                     switch (status.getStatusCode()) {
                         case LocationSettingsStatusCodes.SUCCESS:
                             Log.i(TAG, "All location settings are satisfied.");
-                            // initialize location component wrapper
-                            if (mapboxMap != null) {
-                                mapboxLocationComponentWrapper.init(mapboxMap, getContext());
```
Comment (vincent-karuri) [inline library/src/main/java/io/ona/kujaku/views/KujakuMapView.java]: Why is this removed here?

File: library/src/test/java/io/ona/kujaku/layers/ArrowLineLayerTest.java (hunk 3 above)
Comment (ekigamba) [reply library/src/test/java/io/ona/kujaku/layers/ArrowLineLayerTest.java]: Cool, there are many conventions on naming tests? Should we mix the conventions in one repository? The mostly used convention is **methodWhenGivenShouldDo** or **methodShouldDoWhenGiven** applied appropriately depending on the test. In this case, I probably forgot to include the implicit return expected

File: library/src/test/java/io/ona/kujaku/layers/BoundaryLayerTest.java (hunk 4 above)
Comment (ekigamba) [reply library/src/test/java/io/ona/kujaku/layers/BoundaryLayerTest.java]: Same as above

File: library/src/main/java/io/ona/kujaku/views/KujakuMapView.java (hunk 2 above)
Comment (ekigamba) [reply library/src/main/java/io/ona/kujaku/views/KujakuMapView.java]: OK, does this code need the permissions? The assumption is that this code is not supposed to use it's own location engine. Otherwise that means we have two location engines operating at the same time

File: library/src/main/java/io/ona/kujaku/views/KujakuMapView.java (hunk 5 above)
Comment (ekigamba) [reply library/src/main/java/io/ona/kujaku/views/KujakuMapView.java]: From the code on the lines below this, the default return is `false`. Do we have a reason to change this to `true` instead of `false` from the conditionals

File: library/src/main/java/io/ona/kujaku/views/KujakuMapView.java (hunk 6 above)
Comment (ekigamba) [reply library/src/main/java/io/ona/kujaku/views/KujakuMapView.java]: LayerIds will never be null. The correct annotation to enforce that is in the interface for the `KujakuLayer` and all implementations provide non-null results

File: library/src/main/java/io/ona/kujaku/views/KujakuMapView.java (hunk 8 above)
Comment (ekigamba) [reply library/src/main/java/io/ona/kujaku/views/KujakuMapView.java]: It does not depend on the location permissions. The decision on whether to return this back is going to be informed from this discussion https://github.com/onaio/kujaku/pull/274#discussion_r263392596

File: library/src/main/java/io/ona/kujaku/views/KujakuMapView.java (hunk 6 above)
Comment (vincent-karuri) [reply library/src/main/java/io/ona/kujaku/views/KujakuMapView.java]: OK for this case.

But keep in mind that annotations are merely informational, they don't necessarily provide validation.

No change required here.

File: library/src/main/java/io/ona/kujaku/views/KujakuMapView.java (hunk 8 above)
Comment (vincent-karuri) [reply library/src/main/java/io/ona/kujaku/views/KujakuMapView.java]: OK, moving to https://github.com/onaio/kujaku/pull/274#discussion_r263392596.

File: library/src/test/java/io/ona/kujaku/layers/ArrowLineLayerTest.java (hunk 3 above)
Comment (vincent-karuri) [reply library/src/test/java/io/ona/kujaku/layers/ArrowLineLayerTest.java]: `testMethodShouldDoWhenGiven` works.

File: library/src/test/java/io/ona/kujaku/layers/BoundaryLayerTest.java (hunk 4 above)
Comment (vincent-karuri) [reply library/src/test/java/io/ona/kujaku/layers/BoundaryLayerTest.java]: Refer to https://github.com/onaio/kujaku/pull/274#discussion_r265145397.

File: library/src/main/java/io/ona/kujaku/views/KujakuMapView.java (hunk 2 above)
Comment (vincent-karuri) [reply library/src/main/java/io/ona/kujaku/views/KujakuMapView.java]: I'm not certain why location permissions are requested and therefore chose to leave it that way.

Are you certain it's not needed? Any references?

File: library/src/main/java/io/ona/kujaku/views/KujakuMapView.java (hunk 5 above)
Comment (vincent-karuri) [reply library/src/main/java/io/ona/kujaku/views/KujakuMapView.java]: Returning false by default is OK.

However, I'm not sure if the method should return false if `!mapboxMap.getStyle().isFullyLoaded()`. The layer could still have been added, no?

Comment (ekigamba) [reply library/src/main/java/io/ona/kujaku/views/KujakuMapView.java]: The method is checking to see if the layers that make up the `KujakuLayer` are added. If not but the style is loaded, then the layers aren't added

File: library/src/main/java/io/ona/kujaku/views/KujakuMapView.java (hunk 2 above)
Comment (ekigamba) [reply library/src/main/java/io/ona/kujaku/views/KujakuMapView.java]: Yea, it's not needed:

1. There are no operations that require the location permissions for the location component if you are not using the default location engine. In this case a `null` location engine
2. The current code on master uses the default location engine. If you remove the location check before initialising, you will see an exception in your logs because it tries to access the location. Once you disable the default location engine, it no longer tries to perform any location related operations
3. This is what happens in the code once you disable the location engine https://github.com/mapbox/mapbox-gl-native/blob/master/platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/location/LocationComponent.java#L242 and 
https://github.com/mapbox/mapbox-gl-native/blob/8cc100bfd56955c629d5da0bc49ce2bdbae52946/platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/location/LocationComponent.java#L353 and 
https://github.com/mapbox/mapbox-gl-native/blob/8cc100bfd56955c629d5da0bc49ce2bdbae52946/platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/location/LocationComponent.java#L424 and 
https://github.com/mapbox/mapbox-gl-native/blob/8cc100bfd56955c629d5da0bc49ce2bdbae52946/platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/location/LocationComponent.java#L242 and finally the interaction with the location engine https://github.com/mapbox/mapbox-gl-native/blob/8cc100bfd56955c629d5da0bc49ce2bdbae52946/platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/location/LocationComponent.java#L872

File: library/src/main/java/io/ona/kujaku/views/KujakuMapView.java (hunk 5 above)
Comment (vincent-karuri) [reply library/src/main/java/io/ona/kujaku/views/KujakuMapView.java]: I understand, but what if the style is not fully loaded but the KujakuLayer is already added in the partially loaded style?

File: library/src/test/java/io/ona/kujaku/layers/ArrowLineLayerTest.java (hunk 3 above)
Comment (vincent-karuri) [reply library/src/test/java/io/ona/kujaku/layers/ArrowLineLayerTest.java]: Please counter-check this.

I think you left out the `ShouldDoWhenGiven`. I guess the point would be to follow one convention across all Kujaku tests?

File: library/src/main/java/io/ona/kujaku/views/KujakuMapView.java (hunk 2 above)
Comment (vincent-karuri) [reply library/src/main/java/io/ona/kujaku/views/KujakuMapView.java]: OK, thanks.

File: library/src/main/java/io/ona/kujaku/views/KujakuMapView.java (hunk 5 above)
Comment (ekigamba) [reply library/src/main/java/io/ona/kujaku/views/KujakuMapView.java]: A layer cannot be added when the style is partially loaded. Mapbox does not recognise such a state 
https://docs.mapbox.com/help/tutorials/first-steps-android-sdk/#configure-your-mapview

```
mapView.getMapAsync(new OnMapReadyCallback() {
@Override
public void onMapReady(@NonNull MapboxMap mapboxMap) {
mapboxMap.setStyle(Style.MAPBOX_STREETS, new Style.OnStyleLoaded() {
      @Override
      public void onStyleLoaded(@NonNull Style style) {
        // Map is set up and the style has loaded. Now you can add data or make other map adjustments
      }
    });
    }
});
```

File: library/src/test/java/io/ona/kujaku/layers/ArrowLineLayerTest.java (hunk 3 above)
Comment (ekigamba) [reply library/src/test/java/io/ona/kujaku/layers/ArrowLineLayerTest.java]: Yea, I decided to use `testGetLayerIds` as you requested initially instead of mixing conventions. a test named `methodShouldDoWhenGiven` feels more easily readable compared to `testMethodShouldDoWhenGiven` and since this method has only a single expected outcome or rather no branching. It has one test for which it just returns (like a getter). At this point it would make sense not to name it as the method and have the `test` prefix for easy readability. Are you OK with this?

File: library/src/main/java/io/ona/kujaku/views/KujakuMapView.java (hunk 5 above)
Comment (vincent-karuri) [reply library/src/main/java/io/ona/kujaku/views/KujakuMapView.java]: OK.

File: library/src/test/java/io/ona/kujaku/layers/ArrowLineLayerTest.java (hunk 3 above)
Comment (vincent-karuri) [reply library/src/test/java/io/ona/kujaku/layers/ArrowLineLayerTest.java]: Please choose one convention and carry it through whether or not there is branching in the method or not.

You can add a comment stating whatever convention you choose to use.
